[kernel 2b/n] Add `ChainstateManager::m_adjusted_time_callback`

dongcarl opened the pull request:
```
This decouples validation.cpp from netaddress.cpp (transitively,
timedata.cpp, and asmap.cpp).

This is important for libbitcoinkernel as:

- There is no reason for the consensus engine to be coupled with
  netaddress, timedata, and asmap
- Users of libbitcoinkernel can now easily supply their own
  std::function that provides the adjusted time.

See the src/Makefile.am changes for some satisfying removals.
```

dongcarl committed:
Add ChainstateManager::m_adjusted_time_callback

This decouples validation.cpp from netaddress.cpp (transitively,
timedata.cpp, and asmap.cpp).

This is important for libbitcoinkernel as:

- There is no reason for the consensus engine to be coupled with
  netaddress, timedata, and asmap
- Users of libbitcoinkernel can now easily supply their own
  std::function that provides the adjusted time.

See the src/Makefile.am changes for some satisfying removals.

dongcarl committed:
Add ChainstateManagerOpts, using as ::Options

[META] Although it seems like we don't need it for just one option,
       we're going to introduce another member to this struct *in the
       next commit*. In future patchsets for libbitcoinkernel decoupling
       it from ArgsManager, even more members will be added here.

MarcoFalke commented:
So even if adjusted time was removed, there'd be a time callback?

dongcarl commented:
@MarcoFalke 

> So even if adjusted time was removed, there'd be a time callback?

If usage of adjusted time in validation was removed, then we'd just remove the `m_adjusted_time_callback` member.

If you're asking about if we'll ever make the normal `GetTime` into a callback that users of libbitcoinkernel can plug into... Possibly! But it's quite low priority. `GetAdjustedTime` is much more of a problem since it relies on netaddress, asmap and friends.

sipa commented:
Why does adjusted time depend on asmap...?

dongcarl commented:
@sipa 

> Why does adjusted time depend on asmap...?

Oh sorry, I misspoke. More precisely: adjusted time depends on netaddress, and linking in netaddress requires linking in asmap.

dongcarl commented:
Pushed 4b5a8a1a03af47d3cb4a7c9020f2a0bf63a563a7 -> 45e24007e8742d53b5b45ec22ebed962ffffe079
- Rebased over merge of #24595
- Added commit that introduces `ChainstateManager::Options` (please see commit message + added in-code comments)

dongcarl commented:
Pushed 45e24007e8742d53b5b45ec22ebed962ffffe079 -> 52804f7ad0da32bc4d7be726e08e0b1e34619acb
- Don't use designated initializers 😢

jamesob reviewed (approved):
ACK 52804f7ad0da32bc4d7be726e08e0b1e34619acb ([`jamesob/ackr/25064.1.dongcarl.kernel_2b_n_add_chainst`](https://github.com/jamesob/bitcoin/tree/ackr/25064.1.dongcarl.kernel_2b_n_add_chainst))

Built, reviewed, unittested locally. Change looks good and is very simple. A few minor points:

- there's a lingering comment in `validation.cpp` that could use updating: https://github.com/jamesob/bitcoin/blob/52804f7ad0da32bc4d7be726e08e0b1e34619acb/src/validation.cpp#L2008
- can't use designated initializers (yet?) because of a VS Code/CI issue:

```
10:42 <_aj_> jamesob: #24531 -- VS 2019 (which CI uses) doesn't support them, and figuring out how to fix that isn't trivial
10:42 <@gribble> https://github.com/bitcoin/bitcoin/issues/24531 | Use designated initializers by MarcoFalke · Pull Request #24531 · bitcoin/bitcoin · GitHub
10:45 <sipsorcery> _aj_: in fairness VS2019 does seem support them it just seems to want to use a massive amount of memory to compile them...
10:47 <_aj_> sipsorcery: doesn't support them with the amount of memory we currently allocate for those CI jobs :-P
```

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 52804f7ad0da32bc4d7be726e08e0b1e34619acb ([`jamesob/ackr/25064.1.dongcarl.kernel_2b_n_add_chainst`](https://github.com/jamesob/bitcoin/tree/ackr/25064.1.dongcarl.kernel_2b_n_add_chainst))

Built, reviewed, unittested locally. Change looks good and is very simple. A few minor points:

- - there's a lingering comment in `validation.cpp` that could use updating: https://github.com/jamesob/bitcoin/blob/52804f7ad0da32bc4d7be726e08e0b1e34619acb/src/validation.cpp#L2008
- - can't use designated initializers (yet?) because of a VS Code/CI issue:

```
10:42 <_aj_> jamesob: #24531 -- VS 2019 (which CI uses) doesn't support them, and figuring out how to fix that isn't trivial
10:42 <@gribble> https://github.com/bitcoin/bitcoin/issues/24531 | Use designated initializers by MarcoFalke · Pull Request #24531 · bitcoin/bitcoin · GitHub
10:45 <sipsorcery> _aj_: in fairness VS2019 does seem support them it just seems to want to use a massive amount of memory to compile them...
10:47 <_aj_> sipsorcery: doesn't support them with the amount of memory we currently allocate for those CI jobs :-P
```
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmKDyWUACgkQepNdrbLE
TwWRFQ//UtNY+1IMPL/xvHWvqwgWm9O9FzXZIswmIX3OVze3X9GFXk1sEb7RejHR
z8jFtQW+g5uh4I4vD85BoRyQPmhD2VSc4TCipRTLOjuzOp67OpgSrTq+my89Myh0
ggEQip73F+XPkQl1EjMF/YEH6W1xnUBtIZXLvrXuKm1W03BK5flfxVzsxKYIoj9p
Rb6zic0AaVnbFSxKFiz3F5QE/Vw5lfBbZGblHNAricmVneEX9UNt5CVhCn0mZ2Ki
Ad2UD3Ugbj/qXoF76QeN44QqMsoi57PxwF8Fs53KsuTjdAJSVnD4Ma3ulaeKAy2C
6RZsjJAFGXAGXVRIWr8vvneZmvXRSdnhYPOQvw5gGJTIJxcADH/pSuyP5kDP/CLA
gfCnq+/eK2zhf6B1u8JzUpEbik6IM0bVYamUShZqVQDOKH4nEf7aoOu7E+j0JbWf
gywqlOD3rpDeUS2o8c4TjRqRCIqWPXgHxRI53nQE6AXQvOnNzdLzmetRqvWlqFLu
SgYv4QE25XkFXjbQMDTE+VQ3Qf9/OiHrtRVdYMCpqEHiN2US9RAwGVOF2zqD3wxK
bcY6j5HtbGP341mhmlD3ERBvpk4qi+L5OQRd1s20LyiTVF4RIodOV9i0P1fLL/qh
nPCyUeEoxDp9xWDR9l7aMqxxxt8thNnBt1uvEqdElmdEDLZKTbk=
=9W+m
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-11-amd64-x86_64-with-glibc2.31

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/local/bin/clang++ CC=/usr/local/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/local/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4.1 -msse4.2 -msse4 -msha  i

Compiler version: Debian clang version 13.0.1-++20220126092033+75e33f71c2da-1~exp1~20220126212112.63
```

</p></details>

dongcarl commented:
Pushed 52804f7ad0da32bc4d7be726e08e0b1e34619acb -> f554689988413479ea9bc561f7efd9d56a5c078f
- Updated lingering comment in `validation.cpp` as pointed out here: https://github.com/bitcoin/bitcoin/pull/25064#pullrequestreview-975757513 (thanks James!)

dongcarl commented:
@jamesob 

> * can't use designated initializers (yet?) because of a VS Code/CI issue:

Yeah that's exactly right. We need to resolve #24531 first before we can use them :-/

JeremyRubin commented on src/bitcoin-chainstate.cpp:
```diff
@@ -70,7 +70,10 @@ int main(int argc, char* argv[])
 
     // SETUP: Chainstate
-    ChainstateManager chainman{chainparams};
+    const ChainstateManager::Options chainman_opts{
```
nit: may be more pragmatic to keep name as 'params' or 'settings' since option made me think std::option.

dongcarl replied:
I don't mind either way! I'm following how `BlockAssembler::Options` and `CConnman::Options` do it, but if others think we should use `Params` or something else for these structs in the future, let me know in this thread, I can change it.

JeremyRubin replied:
fair point -- i personally think config/params/settings are better than options, but consistency is probably a overriding interest.

dongcarl replied:
Going to mark as resolved, feel free to reopen for futher convo!

JeremyRubin commented on src/chainstatemanager_opts.h:
```diff
@@ -17,6 +17,7 @@ class CChainParams;
  */
 struct ChainstateManagerOpts {
     const CChainParams& chainparams;
+    const std::function<int64_t()> adjusted_time_callback{nullptr};
```
good candidate for a not_null https://github.com/bitcoin/bitcoin/issues/24423

given that we assert not null where it is actually used.

dongcarl commented on src/init.cpp:
```diff
@@ -1424,7 +1424,10 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        const ChainstateManager::Options chainman_opts{
+            chainparams,
+        };
+        node.chainman = std::make_unique<ChainstateManager>(chainman_opts);
```
Hmmm, how about having `ChainstateManger::m_chainparams` not be a reference? That way `ChainstateManager` will own it wholly. Let me know if there are other easy solutions I'm missing.

JeremyRubin replied:
that'd be fine to me --i wouldn't have said anything if it didn't seem like the kind of thing leading to weird bugs when it's not clear if the kernel or userspace owns the options passed to it.

dongcarl replied:
Addressed in 211674ce4fa845e872e2af9bfcdfc47c37729c52

MarcoFalke commented on src/chainstatemanager_opts.h:
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2022 The Bitcoin Core developers
```
which folder should this file live in and why?

dongcarl replied:
I don't really much mind where this file lives, so I'm happy to move the file wherever reviewers deem appropriate!

MarcoFalke replied:
I would have assumed it will be located in kernel, based on your comment

> In the longer run, I think eventually (very far off when libbitcoincarl is very minimal) we could put everything that links into libbitcoincarl into the src/kernel/ folder.

https://github.com/bitcoin/bitcoin/pull/25065#discussion_r865910970

dongcarl replied:
Yeah true it would eventually be in `kernel/`, so I can just move it there now!

> libbitcoincarl

💀

dongcarl replied:
Addressed in 43f343811cfb22739f5785a00f5a09d7b599fe04

dongcarl committed:
validation: Have ChainstateManager own m_chainparams

We want m_chainparams to be alive for the duration of
ChainstateManager's lifetime since ChainstateManager's behaviour depends
on m_chainparams.

We could allow for a std::shared_ptr to be passed in as m_chainparams,
but that complicates things further. Given that CChainParams is not an
entity class or struct, we can just copy it and have ChainstateManager
own it.

dongcarl commented:
Pushed f554689988413479ea9bc561f7efd9d56a5c078f -> 211674ce4fa845e872e2af9bfcdfc47c37729c52
- Addressed: https://github.com/bitcoin/bitcoin/pull/25064#discussion_r876010385
- Addressed: https://github.com/bitcoin/bitcoin/pull/25064#discussion_r875076316

dongcarl commented:
Pushed af8e5468b589ea0e43e7d130033bfec23b616717 -> 53494bc7392591336e09d095f1fc38d63d566abf
- Rebased over master to resolve merge conflict

JeremyRubin commented:
ack [53494bc](https://github.com/bitcoin/bitcoin/pull/25064/commits/53494bc7392591336e09d095f1fc38d63d566abf)

InvestorMoney commented on src/validation.h:
```diff
@@ -836,7 +836,7 @@ class ChainstateManager
     CBlockIndex* m_best_invalid GUARDED_BY(::cs_main){nullptr};
 
-    const CChainParams& m_chainparams;
+    const CChainParams m_chainparams;
```
839

InvestorMoney reviewed (commented):
> 
Duplicate of #
